src/history/sqlite_backed.rs

```diff
@@ -0,0 +1,410 @@
+            .map_err(map_sqlite_err)?;
+        db.execute(
+            "
+        create table if not exists history (
+            id integer primary key autoincrement,
+            command_line text not null,
+            start_timestamp integer,
+            session_id integer,
+            hostname text,
+            cwd text,
+            duration_ms integer,
+            exit_status integer,
+            more_info text
+        ) strict;
+        create index if not exists idx_history_time on history(start_timestamp);
+        create index if not exists idx_history_cwd on history(cwd); -- suboptimal for many hosts
+        create index if not exists idx_history_exit_status on history(exit_status);
+        create index if not exists idx_history_cmd on history(command_line);
+        create index if not exists idx_history_cmd on history(session_id);
+        -- todo: better indexes
+        ",
+            params![],
+        )
+        .map_err(map_sqlite_err)?;
+        Ok(SqliteBackedHistory { db })
+    }
+    fn construct_query<'a>(
+        &self,
+        query: &'a SearchQuery,
+        select_expression: &str,
+    ) -> (String, BoxedNamedParams<'a>) {
+        // todo: this whole function could be done with less allocs
+        let (is_asc, asc) = match query.direction {
+            SearchDirection::Forward => (true, "asc"),
+            SearchDirection::Backward => (false, "desc"),
+        };
+        let mut wheres: Vec<&str> = vec![];
+        let mut params: BoxedNamedParams = vec![];
+        if let Some(start) = query.start_time {
+            wheres.push(if is_asc {
+                "timestamp_start > :start_time"
+            } else {
+                "timestamp_start < :start_time"
+            });
+            params.push((":start_time", Box::new(start.timestamp_millis())))
+        }
+        if let Some(end) = query.end_time {
+            wheres.push(if is_asc {
+                ":end_time >= timestamp_start"
+            } else {
+                ":end_time <= timestamp_start"
+            });
+            params.push((":end_time", Box::new(end.timestamp_millis())));
+        }
+        if let Some(start) = query.start_id {
+            wheres.push(if is_asc {
+                "id > :start_id"
+            } else {
+                "id < :start_id"
+            });
+            params.push((":start_id", Box::new(start.0)))
+        }
+        if let Some(end) = query.end_id {
+            wheres.push(if is_asc {
+                ":end_id >= id"
+            } else {
+                ":end_id <= id"
+            });
+            params.push((":end_id", Box::new(end.0)));
+        }
+        let limit = match query.limit {
+            Some(l) => {
+                params.push((":limit", Box::new(l)));
+                "limit :limit"
+            }
+            None => "",
+        };
+        if let Some(command_line) = &query.filter.command_line {
+            // todo: escape %
```

Comment:
Any other forbidden chars/Sanitation things to think about?

Reply:
The `LIKE` operator apparently has _ to mean one char and % to mean any amount of chars. SQLite also supports GLOB syntax which would have `?` and `*` + regexp https://sqlite.org/lang_expr.html#the_like_glob_regexp_match_and_extract_operators